pkg/service/auth_context.go

```diff
@@ -23,99 +31,137 @@ type AuthContext struct {
 	Authorization map[*config.AuthorizationConfig]interface{}
 }
 
-// Evaluate evaluates all steps of the auth pipeline (identity → metadata → policy enforcement)
-func (self *AuthContext) Evaluate() error {
-	self.Identity = make(map[*config.IdentityConfig]interface{})
-	self.Metadata = make(map[*config.MetadataConfig]interface{})
-	self.Authorization = make(map[*config.AuthorizationConfig]interface{})
-
-	// identity (authentication)
-	identityConfigs := self.API.IdentityConfigs
-	for i := range identityConfigs {
-		c := identityConfigs[i]
-		if ret, err := c.Call(self); err != nil {
+type configCallback = func(config common.AuthEvaluator, obj interface{})
+
+// NewAuthContext creates an AuthContext instance
+func NewAuthContext(parentCtx context.Context, req *auth.CheckRequest, apiConfig config.APIConfig) AuthContext {
+
+	return AuthContext{
+		ParentContext: &parentCtx,
+		Request:       req,
+		API:           &apiConfig,
+		Identity:      make(map[*config.IdentityConfig]interface{}),
+		Metadata:      make(map[*config.MetadataConfig]interface{}),
+		Authorization: make(map[*config.AuthorizationConfig]interface{}),
+	}
+
+}
+
+func (authContext *AuthContext) evaluateAuthConfig(ctx context.Context, config common.AuthEvaluator, cb configCallback) error {
+	select {
+	case <-ctx.Done():
+		authCxtLog.Info("Context cancelled objConfig terminating", "config", config)
+		return nil
+	default:
+		if authObj, err := config.Call(authContext); err != nil {
+			authCxtLog.Error(err, "Invalid auth object config")
 			return err
 		} else {
-			self.Identity[&c] = ret
-			break
+			cb(config, authObj)
+			return nil
 		}
 	}
+}
+
+func (authContext *AuthContext) evaluateAuthConfigs(authConfigs []common.AuthEvaluator, cb configCallback) error {
+	errGroup, ctx := errgroup.WithContext(context.Background())
+
+	for _, authConfig := range authConfigs {
+		objConfig := authConfig
+		errGroup.Go(func() error {
+			return authContext.evaluateAuthConfig(ctx, objConfig, cb)
+		})
+	}
+	if err := errGroup.Wait(); err != nil {
+		return err
+	} else {
+		authCxtLog.Info("Successfully fetched all auth objects.", "authConfigs", authConfigs)
+		return nil
+	}
+}
+
+// Evaluate evaluates all steps of the auth pipeline (identity → metadata → policy enforcement)
+func (authContext *AuthContext) Evaluate() error {
+	// identity
+	if err := authContext.evaluateAuthConfigs(authContext.API.IdentityConfigs,
+		func(conf common.AuthEvaluator, authObj interface{}) {
+			v, _ := conf.(*config.IdentityConfig)
+			authContext.Identity[v] = authObj
```

Comment:
Something is broken here. We are no longer storing in the `authContext`'s maps the resulting object returned by each evaluator. Instead, the map seems now to be storing a copy of the config.

From the logs, this is the tail of "[OPA] input" for the example request to `/pets/1` with an OIDC token issued to user John.

```
[...], "metadata":{"uma":{"UserInfo":null,"UMA":{"Endpoint":"http://keycloak:8080/auth/realms/ostia","ClientID":"pets-api","ClientSecret":"523b92b6-625d-4e1e-a313-77e7a8ae4e88"}},"userinfo":{"UserInfo":{"OIDC":"keycloak","ClientID":"authorino","ClientSecret":"2e5246f2-f4ef-4d55-8225-36e725071dee"},"UMA":null}}}}
```

The "metadata" field of the JSON above should be of the form ```{ "userinfo": { the actual claims of the user John returned by keycloak }, "uma": { the actual data of the resource whose ID is 1 returned by keycloak } }```. Adding instead respectively the configs for fecthing the userinfo and the resource data is incorrect.